File: app/src/main/java/com/zygne/confetti/engine/particles/BaseParticle.java
package com.zygne.confetti.engine.particles;


import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;

import com.zygne.confetti.engine.components.DynamicObject2d;
import com.zygne.confetti.engine.math.Rectangle;
import com.zygne.confetti.engine.math.Vector2;
import com.zygne.confetti.engine.util.Randomizer;


/**
 * TODO define class.
 *
 * @author Bardur Thomsen
 * @version 1.0 04/09/2018.
 */

public abstract class BaseParticle extends DynamicObject2d {

    public static final int STATE_ALIVE = 0;        // particle is alive
    public static final int STATE_DEAD = 1;         // particle is dead

    public static final int DEFAULT_LIFETIME = 80;  // default lifetime for a particle
    public static final int MAX_DIMENSION = 8;      // the maximum width or height
    public static final int MAX_SPEED = 8;          // maximum speed (per update)
    public static final float DEFAULT_EXPANSION = 1.01f;

    protected Paint paint;                          //
    protected int state;                            // particle is alive or dead

    private int age;                                // current age of the particle
    private int lifetime;                           // particle dies when it reaches this value
    private float expansion;                        // rate of expansion
    private int alpha = 255;                        // alpha value of color
    private int alphaRate;                          // rate at which alpha value decreases
    private boolean fadeAlpha = true;               // should alpha value be reduced
    private boolean shouldExpand = true;

    public BaseParticle(){}

    public BaseParticle(float x, float y){
        this.position = new Vector2(x, y);
        this.bounds = new Rectangle(x, y, 0, 0);
        this.velocity = new Vector2();
        paint = new Paint();
        paint.setColor(Color.RED);
        init(x, y);
    }

    public BaseParticle(float x, float y, float width, float height) {
        super(x, y, width, height);

        init(x, y);
    }

    private void init(float x, float y) {

        this.position.set(x, y);
        this.state = BaseParticle.STATE_ALIVE;
        int radius = Randomizer.rndInt(1, MAX_DIMENSION);
        this.bounds.set(x, y, radius, radius);
        this.lifetime = DEFAULT_LIFETIME;
        this.age = 0;

        this.velocity = new Vector2((Randomizer.rndDbl(0, MAX_SPEED * 2) - MAX_SPEED),
                (Randomizer.rndDbl(0, MAX_SPEED * 2) - MAX_SPEED));
        this.alpha = 255;
        this.alphaRate = (alpha + 5) / lifetime;
        this.expansion = DEFAULT_EXPANSION;

        // smoothing out the diagonal speed
        if (velocity.distSquared(0, 0) > MAX_SPEED * MAX_SPEED) {
            velocity.multiply(0.7f);
        }

        randomizeColor();
    }

    public int getState() {
        return state;
    }

    public void setState(int state) {
        this.state = state;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public int getLifetime() {
        return lifetime;
    }

    public void setLifetime(int lifetime) {
        this.lifetime = lifetime;
    }

    public void setExpansion(float expansion) {
        this.expansion = expansion;
    }

    public boolean isAlive() {
        return this.state == STATE_ALIVE;
    }

    public boolean isDead() {
        return this.state == STATE_DEAD;
    }

    public void reset(float x, float y) {
        init(x, y);
    }

    @Override
    public void update(float dt) {
        if (this.state != STATE_DEAD) {

            this.age++;
            this.position.sub(velocity);
            this.alpha -= alphaRate;
            this.bounds.multiply(expansion);

            if (this.age >= this.lifetime) {    // reached the end if its life
                this.state = STATE_DEAD;
            }

            if (alpha < 10) {
                this.state = STATE_DEAD;
            }

            paint.setAlpha(alpha);
        }
    }

    private void randomizeColor(){
        paint.setARGB(255, Randomizer.rndInt(0,255), Randomizer.rndInt(0,255), Randomizer.rndInt(0,255));
    }

    @Override
    public void render(Canvas canvas) {
        throw new RuntimeException("BaseParticle.render(Canvas canvas) should only be called by the child class!");
    }
}

File: app/src/main/java/com/zygne/confetti/engine/particles/BaseParticle.kt
package com.zygne.confetti.engine.particles

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import com.zygne.confetti.engine.components.DynamicObject2d
import com.zygne.confetti.engine.math.Rectangle
import com.zygne.confetti.engine.math.Vector2
import com.zygne.confetti.engine.util.Randomizer

abstract class BaseParticle(
    x: Float,
    y: Float,
    val maxSpeed: Int = DEFAULT_MAX_SPEED
) :
    DynamicObject2d(x, y) {
    var paint: Paint
    var state = STATE_ALIVE// particle is alive or dead = 0
    var age = 0// current age of the particle = 0
    var lifetime = DEFAULT_LIFETIME// particle dies when it reaches this value = 0
    var expansion = DEFAULT_EXPANSION // rate of expansion = 0f
    var alpha = 255 // alpha value of color
    var alphaRate = 1 // rate at which alpha value decreases = 0
    val fadeAlpha = true // should alpha value be reduced
    val shouldExpand = true

    init {
        position = Vector2(x, y)
        bounds = Rectangle(x, y, 0f, 0f)
        velocity = Vector2()
        paint = Paint()
        paint.color = Color.RED
        setUp(x, y)
    }

    private fun setUp(x: Float, y: Float) {
        position!![x] = y
        state = STATE_ALIVE
        val radius = Randomizer.randomInteger(1, MAX_DIMENSION)
        bounds!![x, y, radius.toFloat()] = radius.toFloat()
        lifetime = 120f
        age = 0
        velocity = Vector2(
            Randomizer.randomDouble(0f, maxSpeed * 2.toFloat()) - maxSpeed,
            Randomizer.randomDouble(0f, maxSpeed * 2.toFloat()) - maxSpeed
        )
        alpha = 255
        alphaRate = ((alpha + 5) / lifetime).toInt()
        expansion = DEFAULT_EXPANSION
        // smoothing out the diagonal speed
        if (velocity!!.distSquared(0f, 0f) > maxSpeed * maxSpeed) {
            velocity!!.multiply(0.7f)
        }

        randomizeColor()
    }

    val isAlive: Boolean
        get() = state == STATE_ALIVE

    val isDead: Boolean
        get() = state == STATE_DEAD

    open fun reset(x: Float, y: Float) {
        setUp(x, y)
    }

    override fun update(dt: Float) {
        if (state != STATE_DEAD) {
            age++
            position!!.sub(velocity!!)
            alpha -= alphaRate
            bounds!!.multiply(expansion)
            if (age >= lifetime) { // reached the end if its life
                state = STATE_DEAD
            }
            if (alpha < 10) {
                state = STATE_DEAD
            }
            paint.alpha = alpha
        }
    }

    open fun randomizeColor() {
        paint.setARGB(
            255,
            Randomizer.randomInteger(0, 255),
            Randomizer.randomInteger(0, 255),
            Randomizer.randomInteger(0, 255)
        )
    }

    override fun render(canvas: Canvas) {
        throw RuntimeException("BaseParticle.render(Canvas canvas) should only be called by the child class!")
    }

    companion object {
        const val STATE_ALIVE = 0 // particle is alive
        const val STATE_DEAD = 1 // particle is dead
        const val DEFAULT_LIFETIME = 80f // default lifetime for a particle
        const val MAX_DIMENSION = 8 // the maximum width or height
        const val DEFAULT_MAX_SPEED = 8 // maximum speed (per update)
        const val DEFAULT_EXPANSION = 1.01f
    }


}